Type of internal URLs incorrectly recognized as external when starting with protocol

When an internal URL (e.g. to `SITE_DOMAIN`) starts with `http://` (or any other pattern configured in the corresponding setting `EXTERNAL_REGEX_STRING`), it is incorrectly recognized as "external", even though the check for internal URLs is used.
In other words, the test case
```
    def test_internal_check_with_protocol(self):
        # "localhost" is configured as SITE_DOMAIN in settings
        uv = Url(url="http://localhost/public/")
        uv.check_url()
        self.assertEqual(uv.type, 'internal')
```
is failing with:
```
======================================================================
FAIL: test_internal_check_with_protocol (linkcheck.tests.test_linkcheck.InternalCheckTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/timo/job/integreat/django-linkcheck/linkcheck/tests/test_linkcheck.py", line 111, in test_internal_check_with_protocol
    self.assertEqual(uv.type, 'internal')
AssertionError: 'external' != 'internal'
- external
? ^^
+ internal
? ^^
```

--- a/CHANGELOG
+++ b/CHANGELOG
@@ -1,5 +1,6 @@
 Unreleased
 
+* Fix `type` property for internal URLs
 * Fix incorrect message when redirect has broken anchor
   (Timo Ludwig, #128)
 * Breaking change: Treat broken hash anchors as valid

--- a/linkcheck/models.py
+++ b/linkcheck/models.py
@@ -137,8 +137,7 @@ class Url(models.Model):
             elif root_domain.startswith('test.'):
                 root_domain = root_domain[5:]
             internal_exceptions = [
-                'http://' + root_domain, 'http://www.' + root_domain, 'http://test.' + root_domain,
-                'https://' + root_domain, 'https://www.' + root_domain, 'https://test.' + root_domain,
+                f'{protocol}://{sub}{root_domain}' for sub in ['', 'www.', 'test.'] for protocol in ['http', 'https']
             ]
 
         for ex in internal_exceptions:

--- a/linkcheck/models.py
+++ b/linkcheck/models.py
@@ -12,6 +12,7 @@ from django.db import models
 from django.test.client import Client
 from django.test.utils import modify_settings
 from django.utils.encoding import iri_to_uri
+from django.utils.functional import cached_property
 from django.utils.timezone import now
 from requests.exceptions import ConnectionError, ReadTimeout
 from requests.models import REDIRECT_STATI
@@ -69,7 +70,7 @@ class Url(models.Model):
 
     @property
     def type(self):
-        if EXTERNAL_REGEX.match(self.url):
+        if self.external:
             return 'external'
         if self.url.startswith('mailto'):
             return 'mailto'
@@ -81,8 +82,10 @@ class Url(models.Model):
             return 'anchor'
         elif self.url.startswith(MEDIA_PREFIX):
             return 'file'
+        elif self.internal_url.startswith('/'):
+            return 'internal'
         else:
-            return 'unknown'
+            return 'invalid'
 
     @property
     def get_message(self):
@@ -106,31 +109,27 @@ class Url(models.Model):
     def __repr__(self):
         return f"<Url (id: {self.id}, url: {self.url})>"
 
-    @property
-    def external(self):
-        return EXTERNAL_REGEX.match(self.url)
-
-    def check_url(self, check_internal=True, check_external=True, external_recheck_interval=EXTERNAL_RECHECK_INTERVAL):
+    @cached_property
+    def internal_url(self):
         """
-        Return:
-         * True if the link was checked and found valid
-         * False if the link was checked and found invalid
-         * None if the link was not checked
+        Remove current domain from URLs as the test client chokes when trying to test them during a page save
+        They shouldn't generally exist but occasionally slip through
+        If settings.SITE_DOMAINS isn't set then use settings.SITE_DOMAIN
+        but also check for variants: example.org, www.example.org, test.example.org
+
+        In case the URLs is external, `None` is returned.
         """
 
-        self.status = False
+        # If the URL is not external, directly return it without processing
+        if not EXTERNAL_REGEX.match(self.url):
+            return self.url
 
-        # Remove current domain from URLs as the test client chokes when trying to test them during a page save
-        # They shouldn't generally exist but occasionally slip through
-        # If settings.SITE_DOMAINS isn't set then use settings.SITE_DOMAIN
-        # but also check for variants: example.org, www.example.org, test.example.org
-
-        tested_url = self.url  # May receive transformation before being checked
+        # May receive transformation before being checked
+        prepared_url = self.url
 
         internal_exceptions = []
         if SITE_DOMAINS:  # If the setting is present
             internal_exceptions = SITE_DOMAINS
-
         elif getattr(settings, 'SITE_DOMAIN', None):  # try using SITE_DOMAIN
             root_domain = settings.SITE_DOMAIN
             if root_domain.startswith('www.'):
@@ -143,16 +142,45 @@ class Url(models.Model):
             ]
 
         for ex in internal_exceptions:
-            if ex and tested_url.startswith(ex):
-                tested_url = tested_url.replace(ex, '', 1)
+            if ex and prepared_url.startswith(ex):
+                prepared_url = prepared_url.replace(ex, '', 1)
 
-        external = bool(EXTERNAL_REGEX.match(tested_url))
+        # If the URL is still external, return `None`
+        if EXTERNAL_REGEX.match(prepared_url):
+            return None
 
-        if check_internal and not external:
-            self._check_internal(tested_url)
+        logger.debug('Internal URL: %s', prepared_url)
+        return prepared_url
 
-        elif check_external and external:
-            self._check_external(tested_url, external_recheck_interval)
+    @property
+    def internal(self):
+        """
+        Check whether this URL is internal
+        """
+        return self.internal_url is not None
+
+    @property
+    def external(self):
+        """
+        Check whether this URL is external
+        """
+        return not self.internal
+
+    def check_url(self, check_internal=True, check_external=True, external_recheck_interval=EXTERNAL_RECHECK_INTERVAL):
+        """
+        Return:
+         * True if the link was checked and found valid
+         * False if the link was checked and found invalid
+         * None if the link was not checked
+        """
+
+        self.status = False
+
+        if check_internal and self.internal:
+            self._check_internal(self.internal_url)
+
+        elif check_external and self.external:
+            self._check_external(self.url, external_recheck_interval)
 
         else:
             return None

--- a/linkcheck/tests/test_linkcheck.py
+++ b/linkcheck/tests/test_linkcheck.py
@@ -100,6 +100,15 @@ class InternalCheckTestCase(TestCase):
         uv.check_url()
         self.assertEqual(uv.status, True)
         self.assertEqual(uv.message, 'Working internal link')
+        self.assertEqual(uv.type, 'internal')
+
+    def test_internal_check_with_protocol(self):
+        # "localhost" is configured as SITE_DOMAIN in settings
+        uv = Url(url="http://localhost/public/")
+        uv.check_url()
+        self.assertEqual(uv.status, True)
+        self.assertEqual(uv.message, 'Working internal link')
+        self.assertEqual(uv.type, 'internal')
 
     def test_internal_check_broken_internal_link(self):
         uv = Url(url="/broken/internal/link")

